x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/ProfileConfigurations.java

```diff
@@ -8,25 +8,106 @@
+ *
+ * Among the transport profiles, two of them are special: "default" and "_remote_cluster".
+ *
+ * The "default" profile has dedicated settings for both networking (e.g. `transport.tcp.keep_alive`)
+ * and SSL (e.g. `xpack.security.transport.ssl.client_authentication`).
+ * It also accepts networking settings specified with its transport profile name,
+ * e.g. this is valid `transport.profiles.default.tcp.keep_alive` is valid.
+ * But it does *not* allow SSL settings to be specified with its transport profile name,
+ * e.g. this is WRONG `transport.profiles.default.xpack.security.ssl.client_authentication`.
+ *
+ * The "_remote_cluster" profile also has dedicated settings for both networking (e.g. `remote_cluster.tcp.keep_alive`)
+ * and SSL (e.g. `xpack.security.remote_cluster.ssl.client_authentication`).
+ * This profile is completely synthetic in that it does NOT accept either networking or SSL settings
+ * with its transport profile name.
+ * NOTE the "_remote_cluster" profile name is special ONLY when the remote cluster feature is enabled.
+ * If the feature is not enabled, this profile name will be treated just as a normal profile.
+ *
+ * When building SSL configurations for the transport profiles, assuming SSL is enabled,
+ * this class builds a map that contains a configuration for each of the configured transport profiles
+ * (keyed by its name).
+ * The map also contains an entry that has the special key "default" and value being the SSL
+ * configuration for the "default" profile.
+ * If remote cluster is enabled, the map will also contain an entry that has the special key
+ * "_remote_cluster" with the value being the SSL configuration of the synthetic "_remote_cluster" profile.
+ *
+ * NOTE the "_remote_cluster" profile only applies to the new remote cluster model.
+ * The legacy remote cluster model mostly just uses the "default" transport profile.
+ */
 public final class ProfileConfigurations {
 
     private ProfileConfigurations() {}
 
-    public static Map<String, SslConfiguration> get(Settings settings, SSLService sslService, SslConfiguration defaultConfiguration) {
+    /**
+     * Builds SSL configuration for transport profiles.
+     *
+     * @param settings Settings of the ES node
+     * @param sslService For resolving the SSL configuration based on its prefix
+     * @param sslEnabledOnly If true, only include the SSL configuration if SSL is enabled for the profile.
+     *                       If false, SSL configuration is included for a profile regardless whether SSL is actually enabled for it.
+     * @return A map that contains {@link SslConfiguration} for each named transport profile as well
+     *         as an entry for the "default" profile. If the remote_cluster feature is enabled, it also
+     *         contains an entry for the synthetic "_remote_cluster" profile.
+     */
+    public static Map<String, SslConfiguration> get(Settings settings, SSLService sslService, boolean sslEnabledOnly) {
```

Comment:
I find `sslEnabledOnly`  hard to follow due to the intersection with the 2 known profiles. My expectations due to the naming would be to return all profiles with ssl enabled irregardless if a known name or not. As is (assuming i am reading this correctly) - if I pass in true and the remote cluster port ssl is enabled but the transport ssl is not then I only get remote cluster profile ... but if I pass in false for the sslEnabledOnly flag then I still get back the remote cluster profile but also will get back custom and default profiles irregadless if they are enabled or not.  I think this method is trying to be too smart and is complicated by the boolean logic of three independent booleans that conditionally change the output.  I think I am partially confused since I don't understand why we need those conditionals at all .

Reply:
The existing method return profiles regardless whether SSL is enabled for them. It also returns _remote_cluster synthetic profile as long as the feature is enabled (whether SSL is enabled is irrelevant). The similar effect to the `sslEnabledOnly` parameter is done by the caller, i.e.:

For `SecurityNett4Transport`, it is
```java
if (transportSslEnabled) {
    profiles = ProfileConfigurations.get(...);
} else {
    profiles = Map.of();
}
```

For `SecurityServerTransportInterceptor`, it is just
```
profiles = ProfileConfigurations.get(...);
```
because it needs initialize ServerTransportFilter for each filter regardless whether SSL is enabled for it.

With the addition of `remoteClusterSslEnabled`, if we keep the existing pattern, for `SecurityNett4Transport`, the code would be something like
```java
profiles = new HashMap<>();
if (transportSslEnabled) {
    profiles.putAll(ProfileConfigurations.getProfilesForDefaultAndTransportProfiles(...));
} 
if (remoteClusterSslEnabled) {
    assert profiles.containsKey(`_remote_cluster`) == false;
    profiles.putAll(ProfileConfigurations.getProfilesForRemoteCluster(...));
}
profiles = Map.copyOf(profiles);
```

So my choice was between (1) split the `ProfileConfigurations.get` method into 2 or (2) keep the single method but move the condition check inside it. I decided to go with the later because it seems to be less change from the API perspective. It also keep the consistency check (assertions) inside the method. But I am also OK to go with the former approach. Or alternatively, we can let `ProfileConfigurations.get` always builds all profiles and remove entries from the Map in the callers. 

What do you think? Please let me know. Thanks!

Reply:
Just to be clear, my preferred approaches, from most preferred to least, are:
1. Single method with the boolean parameter as is in the PR
2. Split the method into 2 as described above
3. Single method that always build all profiles and let callers remove unwanted entries 

Please let me know if any of these works for you. Or if you have other suggestions. Thanks!